Update for PR #457

randombit opened the pull request:
Addresses some issues in PR #457 I noticed.

randombit commented:
@mgierlings @DanielTBrown I would like your feedback on my modifications to the TLS::Callbacks class in 487832e

With my changes on top of PR 457, users of the 'old' TLS API (ie how callbacks have worked from 1.11.0 to the current release) will not have to make any changes to their code at first. Previously the alert callback changed even for the deprecated compat constructor. Instead applications opt into the API changes by deriving from TLS::Callbacks directly. It also removes some (probably minimal) overhead of carrying empty std::functions, when the application is using the new Callbacks class to set the application actions.

Right now I plan on merging this PR at 7a83b81 sometime in the coming week.

mgierlings commented:
@randombit Thank you, the changes look good to me. I especially like the new names you introduced for the callback methods and the extraction of the legacy callbacks to the Compat_Callbacks class.

randombit commented:
I realized the server still had a channel-specific callback which was the ALPN callback. Moved that to TLS::Callbacks in dd5cda3 and squashed some of the prior commits to reduce log noise.

DanielTBrown commented:
@randombit Thank you, the new callback class looks very nice, and easy to use. If you would like doxygen function parameter comments, I have posted PR 595. The pull request contains no code changes, only comment updates.

SonarQubeAnalysis commented on src/lib/tls/tls_server.cpp:
```diff
@@ -284,441 +310,516 @@ Server::get_peer_cert_chain(const Handshake_State& state) const
+      pending_state.set_expected_next(HANDSHAKE_CCS);
+   }
+
+void Server::session_create(Server_Handshake_State& pending_state,
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) This function/method has 116 lines, which is greater than the 100 lines authorized. Split it into smaller functions. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS138)

SonarQubeAnalysis commented on src/lib/tls/tls_server.cpp:
```diff
@@ -284,441 +310,516 @@ Server::get_peer_cert_chain(const Handshake_State& state) const
+      break;
+
+      default:
+         throw Unexpected_Message("Unknown handshake message received");
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove the code after this "throw". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS1763)

SonarQubeAnalysis commented on src/cli/tls_client.cpp:
```diff
@@ -312,16 +305,19 @@ class TLS_Client final : public Command
             output() << buf[i];
          }
+
+      private:
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this redundant access specifier; it does not change the accessibility level. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS3539)

SonarQubeAnalysis commented on src/lib/tls/tls_client.cpp:
```diff
@@ -42,6 +43,23 @@ class Client_Handshake_State : public Handshake_State
 /*
 * TLS Client Constructor
 */
+Client::Client(Callbacks& callbacks,
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) This function has 9 parameters, which is greater than the 7 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS107)

SonarQubeAnalysis commented on src/lib/tls/tls_blocking.cpp:
```diff
@@ -35,14 +39,15 @@ Blocking_Client::Blocking_Client(read_fn reader,
              offer_version,
              next)
    {
+   printf("hi\n");
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Replace this usage of "printf" by a logger. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS106)

SonarQubeAnalysis commented on src/tests/unit_tls.cpp:
```diff
@@ -444,183 +501,234 @@ Test::Result test_dtls_handshake(Botan::TLS::Protocol_Version offer_version,
+               // If we corrupted a DTLS application message, resend it:
+               if(client->is_active() && corrupt_client_data && server_recv.empty())
+                  client->send(client_sent);
+               if(server->is_active() && corrupt_server_data && client_recv.empty())
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 3 if|for|do|while|switch statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

SonarQubeAnalysis commented on src/tests/unit_tls.cpp:
```diff
@@ -444,183 +501,234 @@ Test::Result test_dtls_handshake(Botan::TLS::Protocol_Version offer_version,
-               {
-               result.test_eq("client recv", client_recv, server_sent);
-               }
+               if(client_recv.size())
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 3 if|for|do|while|switch statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

SonarQubeAnalysis commented on src/tests/unit_tls.cpp:
```diff
@@ -218,164 +223,216 @@ Test::Result test_tls_handshake(Botan::TLS::Protocol_Version offer_version,
 
-            if(client.is_closed() && server.is_closed())
-               break;
+               if(client->is_closed() && server->is_closed())
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 3 if|for|do|while|switch statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

SonarQubeAnalysis commented on src/tests/unit_tls.cpp:
```diff
@@ -161,7 +162,11 @@ std::function<void (const byte[], size_t)> queue_inserter(std::vector<byte>& q)
 
-void print_alert(Botan::TLS::Alert, const byte[], size_t)
+void print_alert(Botan::TLS::Alert)
+   {
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add a nested comment explaining why this method is empty, or complete the implementation. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS1186)

SonarQubeAnalysis commented on src/tests/unit_tls.cpp:
```diff
@@ -218,164 +223,216 @@ Test::Result test_tls_handshake(Botan::TLS::Protocol_Version offer_version,
-               {
-               Botan::SymmetricKey client_key = client.key_material_export("label", "context", 32);
-               Botan::SymmetricKey server_key = server.key_material_export("label", "context", 32);
+               if(server_recv.size() && client_recv.size())
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 3 if|for|do|while|switch statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

SonarQubeAnalysis commented on src/tests/unit_tls.cpp:
```diff
@@ -444,183 +501,234 @@ Test::Result test_dtls_handshake(Botan::TLS::Protocol_Version offer_version,
-               {
-               result.test_eq("server recv", server_recv, client_sent);
-               }
+               if(server_recv.size())
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 3 if|for|do|while|switch statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

SonarQubeAnalysis commented on src/tests/unit_tls.cpp:
```diff
@@ -444,183 +501,234 @@ Test::Result test_dtls_handshake(Botan::TLS::Protocol_Version offer_version,
 
-            if(client.is_closed() && server.is_closed())
-               break;
+               if(client->is_closed() && server->is_closed())
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 3 if|for|do|while|switch statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

SonarQubeAnalysis commented on src/tests/unit_tls.cpp:
```diff
@@ -444,183 +501,234 @@ Test::Result test_dtls_handshake(Botan::TLS::Protocol_Version offer_version,
-               {
-               Botan::SymmetricKey client_key = client.key_material_export("label", "context", 32);
-               Botan::SymmetricKey server_key = server.key_material_export("label", "context", 32);
+               if(server_recv.size() && client_recv.size())
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 3 if|for|do|while|switch statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

SonarQubeAnalysis commented on src/tests/unit_tls.cpp:
```diff
@@ -444,183 +501,234 @@ Test::Result test_dtls_handshake(Botan::TLS::Protocol_Version offer_version,
-               std::swap(c2s_traffic, input);
-
-               if(corrupt_server_data)
+               if(client->is_active() && client_sent.empty())
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 3 if|for|do|while|switch statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

SonarQubeAnalysis commented on src/tests/unit_tls.cpp:
```diff
@@ -218,164 +223,216 @@ Test::Result test_tls_handshake(Botan::TLS::Protocol_Version offer_version,
                   }
-               else
+
+               if(client_recv.size())
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 3 if|for|do|while|switch statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

SonarQubeAnalysis commented on src/tests/unit_tls.cpp:
```diff
@@ -444,183 +501,234 @@ Test::Result test_dtls_handshake(Botan::TLS::Protocol_Version offer_version,
                   }
-               else
+
+               if(s2c_traffic.size() > 0)
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 3 if|for|do|while|switch statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

SonarQubeAnalysis commented on src/tests/unit_tls.cpp:
```diff
@@ -218,164 +223,216 @@ Test::Result test_tls_handshake(Botan::TLS::Protocol_Version offer_version,
-               {
-               result.test_eq("server recv", server_recv, client_sent);
-               }
+               if(server_recv.size())
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 3 if|for|do|while|switch statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

SonarQubeAnalysis commented on src/tests/unit_tls.cpp:
```diff
@@ -218,164 +223,216 @@ Test::Result test_tls_handshake(Botan::TLS::Protocol_Version offer_version,
-               server_sent = unlock(rng.random_vec(s_len));
-               server.send(server_sent);
-               }
+               if(client->is_active() && client_sent.empty())
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 3 if|for|do|while|switch statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

SonarQubeAnalysis commented on src/tests/unit_tls.cpp:
```diff
@@ -444,183 +501,234 @@ Test::Result test_dtls_handshake(Botan::TLS::Protocol_Version offer_version,
                   }
-               else
+
+               if(server->is_active() && server_sent.empty())
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 3 if|for|do|while|switch statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

SonarQubeAnalysis commented on src/tests/unit_tls.cpp:
```diff
@@ -218,164 +223,216 @@ Test::Result test_tls_handshake(Botan::TLS::Protocol_Version offer_version,
-            if(r > 2)
-               {
-               if(client_recv.size() && server_recv.size())
+               if(r > 2)
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 3 if|for|do|while|switch statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

SonarQubeAnalysis commented on src/tests/unit_tls.cpp:
```diff
@@ -444,183 +501,234 @@ Test::Result test_dtls_handshake(Botan::TLS::Protocol_Version offer_version,
-            if(server.is_active() && corrupt_server_data && client_recv.empty())
-               server.send(server_sent);
+               // If we corrupted a DTLS application message, resend it:
+               if(client->is_active() && corrupt_client_data && server_recv.empty())
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 3 if|for|do|while|switch statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS134)

SonarQubeAnalysis commented:
SonarQube analysis reported 22 issues
* ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) 21 major
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [tls_client.cpp#L278](https://github.com/randombit/botan/blob/ee60a29088fc6dd712c1651af1e7f56a26f40d63/src/cli/tls_client.cpp#L278): Remove this unused private function. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=cpp%3AS1144)

mgierlings commented:
@randombit Very nice, thanks for the hard work!
